### PR TITLE
Refactored JsonMappingFactory

### PR DESCRIPTION
**Pull request by jakobbraun**
in preparation for #14

### CODE REVIEW

File: src/main/java/com/exasol/adapter/dynamodb/mapping/NestedTableMappingFactory.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ */
+class NestedTableMappingFactory extends AbstractTableMappingFactory {
+    private final TableMappingDefinition parentTable;
+    private final String containingListsPropertyName;
+    private final DocumentPathExpression.Builder sourcePath;
+
+    /**
+     * Creates an instance of {@link NestedTableMappingFactory}.
+     * 
+     * @param parentTable                 the table mapping the object that list mapped by this table is nested in
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/NestedTableMappingFactory.java:19]: This definition is not very clear to me.

```
the parent table that list this table as nested in table
```

or 

```
the table that map this table as a nested in
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            final TableMappingDefinition.Builder tableBuilder) {
+        for (final String dynamodbPropertyName : definition.keySet()) {
+            final DocumentPathExpression.Builder newBuilder = new DocumentPathExpression.Builder(sourcePath)
+                    .addObjectLookup(dynamodbPropertyName);
+            visitMapping(definition.getJsonObject(dynamodbPropertyName), newBuilder, tableBuilder, dynamodbPropertyName,
+                    false);
+        }
+    }
+
+    protected static interface NestedTableReader {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:101]: No class implements this interface yet, but I guess that is expected at the moment?

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java]: No, it is a used by the Labda Expression in line 76. To clarify that it is a [functional interface](https://www.baeldung.com/java-8-lambda-expressions-tips) I will anotate it with @FunctionalInterface. Even though Baeldug suggests to use Fuctiont<T,R> if possible I prefered this custom interface as it adds more meaning, if the method has a good name. Even more i did just made the interface private to signal it is just used in this class.

File: src/main/java/com/exasol/adapter/dynamodb/mapping/JsonMappingFactory.java [hunk 3]
```diff
@@ -23,36 +21,29 @@
...
      * Creates an instance of {@link JsonMappingFactory}.
      *
      * @param definitionsPath path to the definition. Can either be a {@code .json} file or an directory. If it points
      *                        to a directory, all {@code .json} files are loaded.
-     * @throws IOException            if could not open file
-     * @throws SchemaMappingException if schema mapping invalid
+     * @throws IOException                            if could not open file
+     * @throws ExasolDocumentMappingLanguageException if schema mapping invalid
      */
     public JsonMappingFactory(final File definitionsPath) throws IOException, AdapterException {
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/JsonMappingFactory.java:35]: I added a doc for `AdapterException`. `ExasolDocumentMappingLanguageException` is a runtime exception. Even though it needs no throws declaration i wanted to document it.

File: src/main/java/com/exasol/adapter/dynamodb/mapping/NestedTableMappingFactory.java (hunk 1 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/NestedTableMappingFactory.java]: I tried to clearify it. Not shure if it did help...

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    private String getMappingType(final JsonObject definition) {
+        final Set<String> keys = definition.keySet();
+        if (keys.isEmpty()) {
+            return "";
+        } else if (keys.size() == 1) {
+            return keys.iterator().next();
+        } else {
+            throw new ExasolDocumentMappingLanguageException("Please, define only one mapping for one property.");
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:95]:
```suggestion
            throw new ExasolDocumentMappingLanguageException("Please define only one mapping for one property.");
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java (hunk 2 above)
Comment (morazow) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java]: This is very new to me. Thanks for the clarification!

**Pull request merged**